sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,440 @@
+    }
+
+    // Calculate how many entries can fit
+    let data_len = buffer_len - NUM_ENTRIES_SIZE;
+    if data_len % ENTRY_SIZE != 0 {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    let max_entries = data_len / ENTRY_SIZE;
+    if max_entries > MAX_ENTRIES {
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    if let Some(declared) = declared_entries {
+        if declared > max_entries {
+            return Err(ProgramError::InvalidArgument);
+        }
+        return Ok(declared);
+    }
+
+    Ok(max_entries)
+}
+
+/// Validates SlotHashes data format and returns the entry count.
+#[inline]
+fn parse_and_validate_data(data: &[u8]) -> Result<usize, ProgramError> {
+    // Need at least the 8-byte length prefix.
+    if data.len() < NUM_ENTRIES_SIZE {
+        return Err(ProgramError::AccountDataTooSmall);
+    }
+
+    let num_entries = unsafe { read_entry_count_from_bytes_unchecked(data) };
+
+    validate_slothashes_constraints(data.len(), Some(num_entries))
+}
+
+impl SlotHashEntry {
+    /// Returns the slot number as a u64.
+    #[inline(always)]
+    pub fn slot(&self) -> Slot {
+        u64::from_le_bytes(self.slot_le)
+    }
+}
+
+impl<T: Deref<Target = [u8]>> SlotHashes<T> {
+    /// Validates that a buffer is properly sized for SlotHashes data.
+    ///
+    /// Checks that the buffer length is 8 + (N * 40) for some N ≤ 512.
+    #[inline(always)]
+    pub(crate) fn validate_buffer_size(buffer_len: usize) -> Result<(), ProgramError> {
+        validate_slothashes_constraints(buffer_len, None)?;
+        Ok(())
+    }
+
+    /// Creates a `SlotHashes` instance from arbitrary data with full validation.
+    ///
+    /// This constructor performs comprehensive validation of the data format
+    /// including length prefix, entry count bounds, and buffer size requirements.
+    /// Does not validate that entries are sorted in descending order.
+    #[inline(always)]
+    pub fn new(data: T) -> Result<Self, ProgramError> {
+        let num_entries = parse_and_validate_data(&data)?;
+        Ok(unsafe { Self::new_unchecked(data, num_entries) })
+    }
+
+    /// Creates a `SlotHashes` instance directly from a data container and entry count.
+    /// Important: provide a valid len. Whether or not len is assumed to be
+    /// the constant 20_488 (512 entries) is up to caller.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries, where entries are sorted in descending order by slot).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    ///
+    #[inline(always)]
+    pub unsafe fn new_unchecked(data: T, len: usize) -> Self {
```

Comment:
Then here we could have:
```rust
unsafe fn new_unchecked(data: T) -> Self {
    let len = u64::from_le_bytes(unsafe {
        *(data.as_ptr() as *const _ as *const [u8; NUM_ENTRIES_SIZE])
    });
    let entries = unsafe {
        from_raw_parts(
            data.as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry,
            len as usize,
        )
    };

    SlotHashes { data, entries }
}
```

Reply:
The const pointer is gone because it actually added a couple of CUs (<1%, but still) instead of saving them, even in cases of multiple re-derivations of the slice.

`len` is still an open discussion: 512 slot hashes is golden on mainnet, so it's a bit of wasted compute to calculate otherwise. But to serve other contexts, we can let the caller determine whether they want to send in some other length. If they don't want to bother with this at all, they can use `new` instead.